lib/segment/src/types.rs

```diff
@@ -1292,38 +1298,128 @@ pub struct GeoRadius {
 }
 
 impl GeoRadius {
-    pub fn check_point(&self, lon: f64, lat: f64) -> bool {
+    pub fn check_point(&self, point: &GeoPoint) -> bool {
         let query_center = Point::new(self.center.lon, self.center.lat);
-        query_center.haversine_distance(&Point::new(lon, lat)) < self.radius
+        query_center.haversine_distance(&Point::new(point.lon, point.lat)) < self.radius
     }
 }
 
+#[derive(Deserialize)]
+pub struct GeoPolygonShadow {
+    pub exterior: GeoLineString,
+    pub interiors: Option<Vec<GeoLineString>>,
+}
+
 /// Geo filter request
 ///
-/// Matches coordinates inside the polygon, defined by the given coordinates in order.
+/// Matches coordinates inside the polygon, defined by exterior and interiors.
 #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq)]
+#[serde(try_from = "GeoPolygonShadow")]
 #[serde(rename_all = "snake_case")]
 pub struct GeoPolygon {
-    /// Ordered list of coordinates representing the vertices of a polygon.
-    pub points: Vec<GeoPoint>,
+    pub exterior: GeoLineString,
+    pub interiors: Vec<GeoLineString>,
+}
+
+pub struct GeoPolygonValidationError {
+    pub msg: String,
+}
+
+impl std::fmt::Display for GeoPolygonValidationError {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(formatter, "{}", self.msg)
+    }
 }
 
 impl GeoPolygon {
-    pub fn check_point(&self, lon: f64, lat: f64) -> bool {
-        let polygon_points: Vec<Coord<f64>> = self
-            .points
+    pub fn validate_line_string(line: &GeoLineString) -> Result<(), GeoPolygonValidationError> {
+        if line.points.len() <= 3 {
+            return Err(GeoPolygonValidationError {
+                msg: format!(
+                    "polygon invalid, the size must be at least 3, got {}",
+                    line.points.len()
+                ),
+            });
+        }
+
+        if line.points.first() != line.points.last() {
+            return Err(GeoPolygonValidationError {
+                msg: String::from("polygon invalid, the first and the last points should be the same to form a closed line"),
```

Comment:
I see that the first and last points should be equal. And I see that `line.points.len()` can be 3. But if `line.points.len() == 3` and `line.points.first() != line.points.last()`, user provided a single line instead of triangle. I guess `line.points.len() == 3` is incorrect check and it should be `>=4`

Reply:
Also, should we have float equality here without any epsilon?

Reply:
changed to epsilon